shared-libs/transitions/src/transitions/index.js

```diff
@@ -96,60 +52,79 @@ const processChange = (change, callback) => {
       });
     })
     .catch(err => {
       logger.error('transitions: fetch failed for %s : %o', change.id, err);
-      return callback();
+      return callback(err);
     });
 };
 
-const deleteReadDocs = change => {
-  // we don't know if the deleted doc was a report or a message so
-  // attempt to delete both
-  const possibleReadDocIds = ['report', 'message'].map(
-    type => `read:${type}:${change.id}`
-  );
+// given a collection of docs this function will:
+// - deep clone the docs, to keep the original version of them safe
+// - add _id properties to the docs that are missing them
+// - hydrate the docs and generate info docs
+// - run loaded transitions over all docs
+// - returns docs, either updated or their original version
+const processDocs = docs => {
+  if (!transitions.length) {
+    return Promise.resolve(docs);
+  }
 
-  return db.allDbs().then(dbs => {
-    const userDbs = dbs.filter(db => db.indexOf('medic-user-') === 0);
-    return userDbs.reduce((p, userDb) => {
-      return p.then(() => {
-        const metaDb = db.get(userDb);
-        return metaDb.allDocs({ keys: possibleReadDocIds }).then(results => {
-          const row = results.rows.find(row => !row.error);
-          if (!row) {
-            return;
-          }
+  let changes;
+  const clonedDocs = JSON.parse(JSON.stringify(docs)),
+        idsByIdx = [];
+  // keep a way to link the changed doc to the original doc
+  // there may be no reliable way to uniquely identify any of these docs
+  clonedDocs.forEach((doc, idx) => {
+    doc._id = doc._id || uuid();
+    idsByIdx[idx] = doc._id;
+  });
 
-          return metaDb.remove(row.id, row.value.rev).catch(err => {
-            // ignore 404s or 409s - the doc was probably deleted client side already
-            if (err && err.status !== 404 && err.status !== 409) {
-              throw err;
+  return lineage
+    .hydrateDocs(clonedDocs)
+    .then(hydratedDocs => {
+      changes = hydratedDocs.map(doc => ({ id: doc._id, doc: doc, seq: null }));
+      return infodoc.bulkGet(changes);
+    })
+    .then(infoDocs => {
+      changes.forEach(change => {
+        change.info = infoDocs.find(infoDoc => infoDoc.doc_id === change.id);
+      });
+      return infodoc.bulkUpdate(infoDocs);
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        const operations = changes.map(change => callback => {
+          module.exports.applyTransitions(change, (err, changed) => {
+            if (!err && changed) {
+              return callback(null, change.doc);
             }
+
+            // doc was not changed by any transition, so we return the original doc
+            const idx = idsByIdx.findIndex(id => id === change.id);
+            return err ? callback(err) : callback(null, docs[idx]);
           });
         });
+        async.series(operations, (err, results) => {
+          return err ? reject(err) : resolve(results);
+        });
```

Comment:
Just thought about this.. do you guys think I should delete the just-created info-docs for docs that are just echo-ed back?

Reply:
I think this is fine for now - deleting a doc is probably worse for performance than just leaving it there forever...

I think eventually we want the transitions function to return an array of docs to save, so all the docs are saved once all transitions have executed, but that would need careful consideration...

Reply:
I would very much agree with the transaction-like approach.

Reply:
If the transition returns "not changed" or has an error, I return the untouched version of the document that I received, which will not even have an `_id` property. 
The reason is that some transitions add errors on the docs, but still return as "not changed". 
The doc will receive another `_id` upon save, so the info-doc we just created is useless. 
I guess I could only copy the `_id` property over in this case, so it wouldn't be useless?

Reply:
Might as well right? The upside I presume being that when these docs are written Sentinel won't re-run those transactions over them. And the main downside is that we don't generate "smart" UUIDs that CouchDB likes, but honestly we don't do that for 90% of our data right now anyway (cause of PouchDB), and if we really wanted to we could replace the uuid generation used here with a cache that [sources UUIDs from CouchDB](https://docs.couchdb.org/en/stable/api/server/common.html#get--_uuids) if we considered it an issue.

Reply:
Sentinel would run all transitions over that doc either way.

Reply:
That's a shame. Maybe not in the future though (like I feel like we _should_ be able to take advantage of this even if we aren't now)!

Also, it's probably better for Impact to not have orphaned info docs. I may not matter depending on their impl, but it's a bug waiting to happen in any case.

Reply:
This thread is outdated because docs are saved in series within the library immediately after running transitions, same as old sentinel.